File: ros2-bridge/tests/test_resync.py
"""Resync-bound tests.

ADR-0002 commits to a documented resync bound: after arbitrary
corruption ends, the parser must lock onto the next valid frame within
``MAX_PAYLOAD + 10`` bytes. These tests are the regression gate.
"""

from __future__ import annotations

from hypothesis import given, settings
from hypothesis import strategies as st

from rosserial2.codec import (
    MAX_PAYLOAD,
    OVERHEAD,
    Frame,
    FrameParser,
    encode_frame,
)

frame_st = st.builds(
    lambda seq, msg_id, payload: (seq, msg_id, payload),
    seq=st.integers(min_value=0, max_value=0xFF),
    msg_id=st.integers(min_value=0, max_value=0xFF),
    payload=st.binary(min_size=0, max_size=MAX_PAYLOAD),
)


@given(
    junk=st.binary(min_size=0, max_size=2048),
    args=frame_st,
)
def test_arbitrary_prefix_then_valid(junk: bytes, args: tuple[int, int, bytes]) -> None:
    """A valid frame after arbitrary junk must still decode."""
    seq, msg_id, payload = args
    encoded = encode_frame(seq, msg_id, payload)
    parser = FrameParser()
    out = parser.feed(junk + encoded)
    assert Frame(seq=seq, msg_id=msg_id, payload=payload) in out


@given(
    pre=st.binary(min_size=0, max_size=512),
    args1=frame_st,
    middle_junk=st.binary(min_size=0, max_size=512),
    args2=frame_st,
)
@settings(max_examples=300)
def test_two_frames_with_junk_between(
    pre: bytes,
    args1: tuple[int, int, bytes],
    middle_junk: bytes,
    args2: tuple[int, int, bytes],
) -> None:
    s1, m1, p1 = args1
    s2, m2, p2 = args2
    enc1 = encode_frame(s1, m1, p1)
    enc2 = encode_frame(s2, m2, p2)
    parser = FrameParser()
    out = parser.feed(pre + enc1 + middle_junk + enc2)
    assert Frame(seq=s1, msg_id=m1, payload=p1) in out
    assert Frame(seq=s2, msg_id=m2, payload=p2) in out


@given(args=frame_st, drop_at=st.integers(min_value=0))
@settings(max_examples=200)
def test_truncated_frame_does_not_break_following(
    args: tuple[int, int, bytes],
    drop_at: int,
) -> None:
    """A truncated frame must not poison the parser for the next one.

    The parser is streaming: a truncated frame whose claimed length is
    inflated may legitimately wait for more bytes. We feed enough
    trailing data (payload-of-zeros pad) for the rewind path to
    rediscover the next sync within the documented MAX_PAYLOAD + OVERHEAD
    resync bound.
    """
    seq, msg_id, payload = args
    enc = encode_frame(seq, msg_id, payload)
    cut = drop_at % len(enc)
    truncated = enc[:cut]
    follow = encode_frame(0, 0, b"after")
    pad = b"\x00" * (MAX_PAYLOAD + OVERHEAD)
    parser = FrameParser()
    out = parser.feed(truncated + follow + pad)
    assert Frame(seq=0, msg_id=0, payload=b"after") in out


@given(
    args=frame_st,
    flip_at=st.integers(min_value=0, max_value=10000),
)
@settings(max_examples=200)
def test_corrupted_frame_recovers_on_next(
    args: tuple[int, int, bytes],
    flip_at: int,
) -> None:
    """Bit-flip in frame N → parser still locks onto frame N+1.

    A flip in the LEN field can inflate the claimed payload length, so
    the parser legitimately waits for those bytes. We pad with enough
    trailing bytes for the rewind path to rediscover sync within the
    documented MAX_PAYLOAD + OVERHEAD resync bound.
    """
    seq, msg_id, payload = args
    bad = bytearray(encode_frame(seq, msg_id, payload))
    bit = flip_at % (len(bad) * 8)
    byte_idx, bit_idx = divmod(bit, 8)
    bad[byte_idx] ^= 1 << bit_idx
    follow = encode_frame((seq + 1) & 0xFF, msg_id, b"recovered")
    pad = b"\x00" * (MAX_PAYLOAD + OVERHEAD)
    parser = FrameParser()
    out = parser.feed(bytes(bad) + follow + pad)
    assert Frame(seq=(seq + 1) & 0xFF, msg_id=msg_id, payload=b"recovered") in out


def test_resync_bound_explicit() -> None:
    """After corruption ends, lock within MAX_PAYLOAD + OVERHEAD bytes."""
    # Worst-case junk: MAX_PAYLOAD bytes of 0xAA which look like sync starts.
    junk = b"\xaa" * (MAX_PAYLOAD + OVERHEAD)
    valid = encode_frame(42, 7, b"hello")
    parser = FrameParser()
    out = parser.feed(junk + valid)
    assert Frame(seq=42, msg_id=7, payload=b"hello") in out


def test_payload_containing_sync_bytes() -> None:
    """A payload that legally contains 0xAA 0x55 must not confuse the parser."""
    nasty = b"\xaa\x55" * 100
    enc = encode_frame(1, 1, nasty)
    follow = encode_frame(2, 2, b"next")
    parser = FrameParser()
    out = parser.feed(enc + follow)
    assert Frame(seq=1, msg_id=1, payload=nasty) in out
    assert Frame(seq=2, msg_id=2, payload=b"next") in out


def test_payload_looks_like_inner_frame() -> None:
    """Payload that begins with a plausible-looking inner frame header."""
    inner = encode_frame(99, 99, b"trickster")  # full valid inner frame
    enc = encode_frame(1, 1, inner)  # wrapped as payload
    follow = encode_frame(2, 2, b"after")
    parser = FrameParser()
    out = parser.feed(enc + follow)
    # The outer frame must come out intact, and the next real frame too.
    assert Frame(seq=1, msg_id=1, payload=inner) in out
    assert Frame(seq=2, msg_id=2, payload=b"after") in out

File: esp32-firmware/host_test/tests/test_resync.cpp
#include <catch2/catch_test_macros.hpp>
#include <random>

#include "rosserial2/codec.hpp"
#include "test_helpers.hpp"

using namespace rosserial2;
using test_helpers::Sink;

TEST_CASE("parser locks onto frame after arbitrary junk prefix", "[resync]") {
    Sink sink;
    FrameParser parser;
    const std::vector<std::uint8_t> junk{
        0x00, 0xFF, 0xAA, 0xAB, 0x55, 0x55, 0xAA, 0x55,  // teaser sync — bad LEN coming
        0xFF, 0xFF,  // LEN > MAX_PAYLOAD → reject, rewind
        0x12, 0x34, 0x56,
    };
    const auto frame = test_helpers::encode(99, 1, {'o', 'k'});
    std::vector<std::uint8_t> stream = junk;
    stream.insert(stream.end(), frame.begin(), frame.end());
    parser.feed(stream.data(), stream.size(), test_helpers::capture_cb, &sink);
    REQUIRE(sink.frames.size() == 1);
    REQUIRE(sink.frames[0].seq == 99);
    REQUIRE(sink.frames[0].payload == std::vector<std::uint8_t>{'o', 'k'});
}

TEST_CASE("payload containing AA 55 does not confuse the parser", "[resync]") {
    Sink sink;
    FrameParser parser;
    std::vector<std::uint8_t> nasty(200);
    for (std::size_t i = 0; i + 1 < nasty.size(); i += 2) {
        nasty[i] = 0xAA;
        nasty[i + 1] = 0x55;
    }
    const auto f1 = test_helpers::encode(1, 1, nasty);
    const auto f2 = test_helpers::encode(2, 2, {'n', 'e', 'x', 't'});
    std::vector<std::uint8_t> stream = f1;
    stream.insert(stream.end(), f2.begin(), f2.end());
    parser.feed(stream.data(), stream.size(), test_helpers::capture_cb, &sink);
    REQUIRE(sink.frames.size() == 2);
    REQUIRE(sink.frames[0].payload == nasty);
    REQUIRE(sink.frames[1].payload == std::vector<std::uint8_t>{'n', 'e', 'x', 't'});
}

TEST_CASE("payload that looks like an inner frame is preserved", "[resync]") {
    Sink sink;
    FrameParser parser;
    const auto inner = test_helpers::encode(99, 99, {'t', 'r', 'i', 'c', 'k'});
    const auto outer = test_helpers::encode(1, 1, inner);
    const auto follow = test_helpers::encode(2, 2, {'a', 'f', 't', 'e', 'r'});
    std::vector<std::uint8_t> stream = outer;
    stream.insert(stream.end(), follow.begin(), follow.end());
    parser.feed(stream.data(), stream.size(), test_helpers::capture_cb, &sink);
    REQUIRE(sink.frames.size() == 2);
    REQUIRE(sink.frames[0].payload == inner);
    REQUIRE(sink.frames[1].payload == std::vector<std::uint8_t>{'a', 'f', 't', 'e', 'r'});
}

TEST_CASE("truncated frame does not poison the next one", "[resync]") {
    Sink sink;
    FrameParser parser;
    const auto enc = test_helpers::encode(5, 5, std::vector<std::uint8_t>(20, 0x42));
    const auto follow = test_helpers::encode(0, 0, {'a', 'f', 't', 'e', 'r'});
    const std::vector<std::uint8_t> pad(MAX_PAYLOAD + OVERHEAD, 0);
    for (std::size_t cut = 0; cut < enc.size(); ++cut) {
        sink.frames.clear();
        parser.reset();
        std::vector<std::uint8_t> stream(enc.begin(), enc.begin() + static_cast<long>(cut));
        stream.insert(stream.end(), follow.begin(), follow.end());
        stream.insert(stream.end(), pad.begin(), pad.end());
        parser.feed(stream.data(), stream.size(), test_helpers::capture_cb, &sink);
        bool found = false;
        for (const auto& f : sink.frames) {
            if (f.seq == 0 && f.msg_id == 0 &&
                f.payload == std::vector<std::uint8_t>{'a', 'f', 't', 'e', 'r'}) {
                found = true;
                break;
            }
        }
        REQUIRE(found);
    }
}

TEST_CASE("explicit resync bound: MAX_PAYLOAD + OVERHEAD bytes of junk", "[resync]") {
    Sink sink;
    FrameParser parser;
    std::vector<std::uint8_t> junk(MAX_PAYLOAD + OVERHEAD, 0xAA);  // worst case
    const auto good = test_helpers::encode(42, 7, {'h', 'e', 'l', 'l', 'o'});
    std::vector<std::uint8_t> stream = junk;
    stream.insert(stream.end(), good.begin(), good.end());
    parser.feed(stream.data(), stream.size(), test_helpers::capture_cb, &sink);
    REQUIRE(sink.frames.size() == 1);
    REQUIRE(sink.frames[0].seq == 42);
    REQUIRE(sink.frames[0].msg_id == 7);
    REQUIRE(sink.frames[0].payload == std::vector<std::uint8_t>{'h', 'e', 'l', 'l', 'o'});
}

TEST_CASE("bit-flip in frame N — frame N+1 still decodes", "[resync][bitflip]") {
    std::mt19937 rng(0xCAFEFEEDu);
    std::uniform_int_distribution<int> len_d(0, 32);
    std::uniform_int_distribution<int> byte_d(0, 255);
    const std::vector<std::uint8_t> pad(MAX_PAYLOAD + OVERHEAD, 0);
    for (int i = 0; i < 200; ++i) {
        std::vector<std::uint8_t> p(static_cast<std::size_t>(len_d(rng)));
        for (auto& b : p) b = static_cast<std::uint8_t>(byte_d(rng));
        auto bad = test_helpers::encode(0, 0, p);
        std::uniform_int_distribution<std::size_t> bit_d(0, bad.size() * 8 - 1);
        const auto bit = bit_d(rng);
        bad[bit / 8] ^= static_cast<std::uint8_t>(1u << (bit % 8));
        const auto good = test_helpers::encode(1, 0, {'r', 'e', 'c', 'o', 'v'});
        std::vector<std::uint8_t> stream = bad;
        stream.insert(stream.end(), good.begin(), good.end());
        stream.insert(stream.end(), pad.begin(), pad.end());
        Sink sink;
        FrameParser parser;
        parser.feed(stream.data(), stream.size(), test_helpers::capture_cb, &sink);
        bool found = false;
        for (const auto& f : sink.frames) {
            if (f.seq == 1 && f.msg_id == 0 &&
                f.payload == std::vector<std::uint8_t>{'r', 'e', 'c', 'o', 'v'}) {
                found = true;
                break;
            }
        }
        REQUIRE(found);
    }
}
